Ignition Gazebo Architecture intro (#142)
ignitionrobotics/docs: architecture.md

# Ignition Gazebo Architecture

Ignition Gazebo is the application entry-point of Ignition architecture, generally encompassing the use of all other Ignition libraries.
It is a library itself, but its use case described in this document is as an executable.
As an executable, it runs a simulation by launching two processes: the backend server process and frontend client process.

The core of Ignition Gazebo is the server-client communication and the various plugins it loads to facilitate the simulation.
Plugins are loaded at runtime to create a link between other Ignition libraries and Ignition Gazebo so they can exchange information.
Users can write their own plugins.

> Because they’re loaded at runtime, Ignition does not need to be recompiled to add or remove plugins.

Reviewer: Might be good to clarify that the users' plugins are compiled into library files. Right now it reads like a missing link, between the users writing the plugins, and the plugins magically getting loaded at runtime, but it's unclear how they turned from code into something loadable.

Author: I was wondering myself what ties the plugins to the libraries. Will definitely be adding this!

> There are other types of plugins in the scope of Ignition; however, this use case only concerns Ignition Gazebo server and GUI plugins.

Reviewer: Is "this use case" referring to the "other types of plugins," or the type described earlier in the paragraph?

Author: Hmm.. I guess what I meant was "In the scope of this article", referring back to line 4 where I said "its [Ignition Gazebo's] use case described in this document is as an executable." 

Or possibly "the use case of Ignition Gazebo as an executable" makes more sense for line 11. I'm not sure if either of those re-phrasings is accurate enough to state.

Without plugins, Ignition libraries are not “aware” of each other and don’t possess the ability to interact.

> For example, Ignition Physics is independent of Ignition Gazebo, and vice versa.

Reviewer: It would be good to add as the first sentence of the paragraph, emphasizing that the libraries are independent and modular. It's a selling point.

So Ignition Gazebo has a physics plugin that uses Ignition Physics as a library and incorporates Gazebo specifics, to enable Physics to be a system running in the simulation loop.
This is the case for all libraries, whether their corresponding plugins affect the frontend or backend processes.

The Ignition Physics library is only used in the physics plugin, not in other plugins nor in the core of Ignition Gazebo.
Some libraries only have one corresponding plugin, or one in each process (frontend and backend).
However, some libraries are used by all plugins, like Ignition Common which provides common functionality like logging, manipulating strings, file system interaction, etc., to all plugins.

> Other such libraries include Ignition Plugin, Ignition Math, SDFormat, and more.

Reviewer: Would be nice to have links for each of these libraries mentioned, static ones independent of version numbers like https://ignitionrobotics.org/libs/math , but not essential.

There are certain plugins in both the frontend and backend processes that are loaded by default in every version of Ignition.
Many other plugins are optional.
One common optional plugin is the sensors plugin.

> Both optional and default plugins can be removed or added at any time; Ignition Gazebo will continue to run with limited functionality.

Reviewer: It would be great if we have a version-independent URL to a demo example, but I don't know if we already have one. Just a nice to have for the future.

Reviewer 2: That's a great idea, a lot of people (like me) learn better from demos. There are a couple of hands-on guides that we can link to here. They're versioned though:

* [Configuring the GUI](https://ignitionrobotics.org/api/gazebo/4.3/gui_config.html)
* [Configuring the server](https://ignitionrobotics.org/api/gazebo/4.3/server_config.html)

![Ignition Gazebo architecture diagram]()

## Backend server process

> Ignition gazebo is responsible for loading plugins in the backend, referred to as systems.

Reviewer: ```suggestion
Ignition Gazebo is responsible for loading plugins in the backend, referred to as systems.
```

Systems act on entities and components of those entities.
An entity is anything in the simulation scene (a model, link, light, actor, etc.), and components are their characteristics (pose, name, geometry, etc.).
Server plugins have access to all entities and what components they have, and make modifications to those components.
For example, a user could write a system that applies force to an entity by setting a component on that entity, and the physics system would pick that up and react by applying the force to make the entity actually move.
The modification of entities and components is how server plugins communicate with each other.
Libraries, however, don’t know about the entities and components.

All of the systems in the backend create a loop, where some systems are proposing changes to entities and their components, and other systems are handling and applying those changes.
This is called the “Simulation loop”.

Physics, user commands, and scene broadcaster are all systems launched by default in the backend.
As mentioned earlier, however, even default systems can be added to or removed from this loop.

> For any other functionality, like sensor data processing for example, a new system would have to be added.

Reviewer: Maybe "new" -> "additional"? What I'm trying to say is, "new" could sound like, we don't have such systems yet, and someone will have to make a new one. But we have a bunch non-default ones.

> Plugins that generate sensor data and utilize the Ignition Sensors and Ignition Rendering libraries would have to be added to the backend process.

Reviewer: For these two sentences, maybe "new" -> "additional"? Or reword?
Currently, it could sound like, we haven't developed such systems yet, and someone will have to write them. But we have a bunch non-default ones ready to be added.

The visualization of that data, however, is left up to the client side plugins.

## Communication process

> Synchronization between processes is performed by scene broadcaster, a server-side Ignition Gazebo plugin that uses Ignition Transport and Ignition Messages libraries to send messages from the server to the client.

Reviewer: For this sentence, I think it definitely helps to link to Transport and Messages so those interested can read about the protobuf dependency and get a rough idea of what's under the hood
https://ignitionrobotics.org/libs/transport
https://ignitionrobotics.org/libs/msgs

The scene broadcaster system is responsible for getting the state of the world (all of the entities and their component values) from the ever-changing simulation loop running in the back end, packaging that information into a very compact message, and periodically sending it to the frontend to process.
Omitting this system would mean not being able to visualize data on the frontend, which can be useful for saving computational power.

In addition to state messages sent from the scene broadcaster to the client, the client can also make requests to the server using the user commands system (another backend plugin).
It also utilizes the Ignition Transport and Ignition Messages libraries.
The user command system has services that can be requested by the frontend GUI to insert, delete, move, etc., and will send responses back to the GUI acknowledging the receipt and execution of those requests.

## Frontend client process

The client-side process, essentially the GUI, also comprises multiple plugins, some loaded by default, others that can be added, and all optional.

> All of the frontend plugins use the Ignition GUI library.

Reviewer: https://ignitionrobotics.org/libs/gui

> There are visualization plugins that create the windows of the GUI, add buttons and other interactive features, and a 3D scene plugin that utilizes Ignition Rendering that creates the scene the user sees.

Reviewer: If we want to add links
https://ignitionrobotics.org/libs/gui
https://ignitionrobotics.org/libs/rendering

However, some foundational libraries are used by all plugins, like [Ignition Common](https://ignitionrobotics.org/libs/common) which provides common functionality like logging, manipulating strings, file system interaction, etc., to all plugins.
Other such libraries include [Ignition Plugin](https://ignitionrobotics.org/libs/plugin), [Ignition Math](https://ignitionrobotics.org/libs/math), [SDFormat](https://ignitionrobotics.org/libs/sdformat), and more.

There are certain plugins in both the frontend and backend processes that are loaded by default in every version of Ignition.
Many other plugins are optional.
One common optional plugin is the sensors plugin.
Both optional and default plugins can be removed or added at any time; Ignition Gazebo will continue to run with limited functionality.
These demos on [Server Configuration](https://ignitionrobotics.org/api/gazebo/4.3/server_config.html) and [GUI Configuration](https://ignitionrobotics.org/api/gazebo/4.3/gui_config.html) showcase that functionality.

The simulation process is depicted in the diagram below, and further explained in the Backend and Frontend process sections that follow.

> ![Ignition Gazebo architecture diagram](images/IgnitonGazeboArchicteture.png)

Reviewer 2: This image is great! About time we had something like this! I love the little circles representing common Ignition libraries.

I just have some minor comments:

* The arrow that says "Packaged Messages" could say "Serialized ECM state" or something to that effect.
* How about using different arrows for the ones that represent Ignition Transport (the ones going across processes) and the ones connecting systems to the ECMs?
* Just a personal preference, but how about using a sans serif font? I think they're slightly cleaner. 

Finally, would it be possible for you to save this image as SVG and upload it as well? That should make it easier to edit it in the future. Thanks!

## Backend server process

Ignition Gazebo is responsible for loading plugins in the backend, referred to as systems.
The server runs an entity-component system architecture (see [Ignition Gazebo terminology](https://ignitionrobotics.org/api/gazebo/4.2/terminology.html)).
The backend will usually have multiple systems responsible for everything in the simulation – computing physics, recording logs, receiving user commands, etc.

Systems act on entities and components of those entities.
An entity is anything in the simulation scene (a model, link, light, actor, etc.), and components are their characteristics (pose, name, geometry, etc.).
Server plugins have access to all entities and what components they have, and make modifications to those components.
For example, a user could write a system that applies force to an entity by setting a component on that entity, and the physics system would pick that up and react by applying the force to make the entity actually move.
The modification of entities and components is how server plugins communicate with each other.

> Other Ignition Libraries, however, don’t know about the entities and components.

Reviewer 3: I think the last sentence of this paragraph might be a little confusing with respect to the other sentences/content of this paragraph. We give an example of the physics system (plugin) operating on an entity, which might make readers think that `Ignition Physics` has direct access to the entity component system, but then finish the paragraph by saying that Ignition libraries don't know about entities and components. So, would it be better to instead say something like:
```
The entity component system is self-contained in Ignition Gazebo, and systems act on entities and their components.
Since other Ignition Libraries don't have direct knowledge/access to entities and components, other Ignition Libraries should be used in systems so that the server plugin can share entities (and their components) with these other Libraries.
For example, the physics system shares entities and components with the Ignition Physics library in order to achieve physics effects on entities and components.
```

If the assumptions I made in the re-phrase above are incorrect, please let me know - I am still relatively new to Ignition :upside_down_face: This is just a thought/suggestion I had. Feel free to modify it in whatever way that makes the most sense!

There is a loop running in the backend that runs the systems, where some systems are proposing changes to entities and their components, and other systems are handling and applying those changes.

> This is called the “simulation loop”.

Reviewer 3: I believe the "simulation loop" consists of `PreUpdate`, `Update`, and `PostUpdate`, right? Perhaps we can talk about this here, unless that is a little too technical/detailed for the scope of this tutorial. @chapulina what do you think?

Reviewer 2: @maryaB-osr and I briefly discussed this idea but we thought this may be a good topic for a follow-up tutorial, not to make this one too long.

Reviewer 3: Doing this in a follow-up tutorial is fine with me. I'll mark this as resolved.

The Entity Component Manager (ECM) in the backend provides the functionality for the actual querying and updating of the entities and components.

> Physics, user commands, and scene broadcaster are all systems launched by default in the backend.
> As mentioned earlier, however, even default systems can be added to or removed from this loop.

Reviewer 3: ```suggestion
Physics, user commands, and scene broadcaster are all systems launched by default in the backend.
As mentioned earlier, however, even default systems can be added to or removed from the simulation loop (the [Server Configuration](https://ignitionrobotics.org/api/gazebo/4.3/server_config.html) tutorial describes how to customize default systems).
```

I think it would be good to provide users with more information about adding/removing default systems in case they're curious.

I also replaced "this loop" with "simulation loop" for clarity (users may not recall what "this loop" is since this is a new paragraph).

For any other functionality, like sensor data processing for example, an additional system would have to be loaded.
For example, if you need to generate sensor data that utilizes rendering sensors, you would need to load the sensors system.
The visualization of that data, however, is left up to the client side plugins.

## Communication process

> Any information that crosses the process boundary (whether back-to-front with scene broadcaster or front-to-back with user commands) has to go through [Ignition Transport](https://ignitionrobotics.org/libs/transport), the communication library.

Reviewer 3: I think that this is the first time in the tutorial the term "process boundary" is used. Should we provide a definition?

> Synchronization between processes is performed by scene broadcaster, a server-side Ignition Gazebo plugin that uses Ignition Transport and [Ignition Messages](https://ignitionrobotics.org/libs/msgs) libraries to send messages from the server to the client.

Reviewer 3: ```suggestion
Synchronization between processes is performed by scene broadcaster, a server-side Ignition Gazebo plugin that uses the Ignition Transport and [Ignition Messages](https://ignitionrobotics.org/libs/msgs) libraries to send messages from the server to the client.
```

The messages themselves are provided by Ignition Messages, while the framework for creating the publishers and subscribers that exchange messages is from Ignition Transport.

The scene broadcaster system is responsible for getting the state of the world (all of the entities and their component values) from the ever-changing simulation loop running in the back end, packaging that information into a very compact message, and periodically sending it to the frontend to process.
Omitting the scene broadcaster, while possible, would mean not being able to visualize data on the frontend, which can be useful for saving computational power.

In addition to state messages sent from the scene broadcaster to the client, the client can also make requests to the server using the user commands system (another backend plugin).
It also utilizes the Ignition Transport and Ignition Messages libraries.

> The user command system has services that can be requested by the frontend GUI to insert, delete, move, etc., and will send responses back to the GUI acknowledging the receipt and execution of those requests.

Reviewer 3: Should we say what these "operations" (insert, delete, move, etc.) are being performed on? Are they being performed on models, or is there a chance that they can be performed on other things, too?

Reviewer 2: Models, lights (and eventually actors)

Several non-default plugins are able to send or receive messages to other processes outside of the server-client process.
For example, the server-side sensors plugin publishes messages outside of the server-client process, and the server-side diff-drive plugin receives info from processes other than the client.

> One such process these plugins may communicate with, for example, is ROS.

Reviewer 3: Should we provide a link to `ROS`, in case someone reading this is using Ignition for a non-robotics use case and isn't familiar with ROS?

## Frontend client process

The client-side process, essentially the GUI, also comprises multiple plugins, some loaded by default, others that can be added, and all optional.
All of the frontend plugins use the [Ignition GUI](https://ignitionrobotics.org/libs/gui) library.
The client itself also relies on Ignition GUI.
There are visualization plugins that create the windows of the GUI, add buttons and other interactive features, and a 3D scene plugin that utilizes [Ignition Rendering](https://ignitionrobotics.org/libs/rendering) that creates the scene the user sees.

> Frontend plugins communicate using events.

Reviewer 3: Can we define/describe what an "event" is?

Reviewer 2: Good idea. Maybe something like

> Frontend plugins typically communicate among themselves using [events](https://ignitionrobotics.org/api/gui/4.2/namespaceignition_1_1gui_1_1events.html). Events are similar to messages, but they're processed synchronously. For example, the `Render` event is emitted by a 3D scene from it's rendering thread right before the scene is rendered; this gives other plugins the chance to execute code right at that thread at that moment, which is valuable to edit the 3D scene. Other such events are emitted when the user right-clicks or hovers the scene for example.

They all have access to the entity and component information from the compressed message provided by the scene broadcaster system from the backend.

> The Entity Component Manager on the client side performs the actual message unpacking and distributes the entity and component information to the relevant plugins.

Reviewer 3: In the diagram, we have an arrow that goes from server to client labelled `packed messages`, which points directly to the `unpacked message contents` ECM on the client side. Should we add something to the diagram on the client side that describes this message unpacking process? @chapulina feel free to chime in with your thoughts

Reviewer 2: humm I'm not sure how to add this to the diagram itself. I'm also not sure what other information to provide that can make the process clearer...
